include/seqan3/argument_parser/auxiliary.hpp

```diff
@@ -15,12 +15,211 @@
+
+namespace seqan3::custom
+{
+
+/*!\brief A type that can be specialised to provide customisation point implementations for the seqan3::argument_parser
+ *        such that third party types model may be adapted.
+ * \tparam t The type you wish to specialise for.
+ * \ingroup argument_parser
+ *
+ * \details
+ *
+ * ### Named Enumerations
+ *
+ * In order to use a third party type within the seqan3::argument_parser::add_option or
+ * seqan3::argument_parser::add_positional_option call, you can specialise this struct in the following way:
+ *
+ * \include test/snippet/argument_parser/named_enumeration.cpp
+ *
+ * Please note that by default the `t const`, `t &` and `t const &` specialisations of this class inherit the
+ * specialisation for `t` so you usually only need to provide a specialisation for `t`.
+ *
+ * \note Only use this if you cannot provide respective functions in your namespace. See the tutorial
+ * \ref tutorial_argument_parser for an example of customising a type within your own namespace.
+ */
+template <typename t>
+struct argument_parsing
+{}; // forward
+
+//!\cond
+template <typename t>
+struct argument_parsing<t const> : argument_parsing<t>
+{};
+
+template <typename t>
+struct argument_parsing<t &> : argument_parsing<t>
+{};
+
+template <typename t>
+struct argument_parsing<t const &> : argument_parsing<t>
+{};
+//!\endcond
+
+} // seqan3::custom
+
+namespace seqan3::detail::adl_only
+{
+
+//!\brief Poison-pill overload to prevent non-ADL forms of unqualified lookup.
+template <typename t>
+std::unordered_map<std::string_view, t> enumeration_names(t v) = delete;
+
+/*!\brief Functor definition for seqan3::enumeration_names.
+ *
+ * This struct is templetised because we need the original option_t next to s_option_t that may be wrapped
+ * std::type_identity to be default constructible. We need option_t to be default constructible because the
+ * respective CPO should be callable without a function parameter.
+ */
+template <typename option_t>
+struct enumeration_names_fn
+{
+private:
+    //!\brief `option_t` with cvref removed and possibly wrapped in std::type_identity.
+    using s_option_t = std::conditional_t<std::is_nothrow_default_constructible_v<remove_cvref_t<option_t>> &&
+                                        seqan3::is_constexpr_default_constructible_v<remove_cvref_t<option_t>>,
+                                        remove_cvref_t<option_t>,
+                                        std::type_identity<option_t>>;
+
+    SEQAN3_CPO_IMPL(1, (deferred_type_t<seqan3::custom::argument_parsing<option_t>, decltype(v)>::enumeration_names))
+    SEQAN3_CPO_IMPL(0, (enumeration_names(v))) // ADL
+
+public:
+    //!\brief Operator definition.
+    template <typename dummy = int> // need to make this a template to enforce deferred initialisation
+    //!\cond
+        requires requires
+        {
+            { impl(priority_tag<1>{}, s_option_t{}, dummy{}) } -> std::unordered_map<std::string_view,
```

Comment:
```suggestion
            std::same_as<decltype(impl(priority_tag<1>{}, s_option_t{}, dummy{})),
                         std::unordered_map<std::string_view, remove_cvref_t<option_t>>>;
```

Please use syntax compatible with C++20.